src/day9.scala

```diff
@@ -0,0 +1,112 @@
+// using scala 3.1.0
+
+package day9
+
+import scala.collection.immutable.Queue
+import scala.util.Using
+import scala.io.Source
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day9"))(_.mkString)
+
+type Height = Int
+case class Position(x: Int, y: Int)
+
+class Heightmap(val width: Int, val height: Int, private val data: Array[Height]):
+
+  def apply(coord: Position): Height = data(coord.y * width + coord.x)
+
+  def neighborsOf(coord: Position): Iterator[(Position, Height)] =
+    val Position(x, y) = coord
+    val idx = coord.y * width + coord.x
+    Iterator(
+      Option.when(x > 0)(Position(x - 1, y) -> data(idx - 1)),
+      Option.when(x < width - 1)(Position(x + 1, y) -> data(idx + 1)),
+      Option.when(y > 0)(Position(x, y - 1) -> data(idx - width)),
+      Option.when(y < height - 1)(Position(x, y + 1) -> data(idx + width))
+    ).flatMap(Iterator.from)
+ 
+  def lowPointsPositions: Iterator[Position] =
+    Iterator.range(0, data.length).map { idx =>
+      val x = idx % width
+      val y = idx / width
+      val coord = Position(x, y)
+      (
+        data(idx),
+        coord,
+        this.neighborsOf(coord).map(_._2)
+      )
+    }.collect {
+      case (value, coords, neighbors) if neighbors.forall(value < _) => 
+        coords
+    }
+end Heightmap
+
+ 
+object Heightmap:
+  def fromString(raw: String): Heightmap =
+    val width = raw.linesIterator.next.length
+    val height = raw.linesIterator.length
+    val data = raw.iterator.filter(_.isDigit).map(_.asDigit).toArray
+
+    Heightmap(width, height, data)
+end Heightmap
+
+
+def drawGrid(height: Int, width: Int, data: (Position, Height)*) =
+  val provided: Map[Position, Height] = data.toMap
+  var x = 0
+  var y = 0
+  val result = StringBuilder()
+  while (y < height) do
+    while (x < width) do
+      val value = provided.getOrElse(Position(x, y), -1)
+      result.append(if value >= 0 then value else '-')
+      x += 1
+    result.append('\n')
+    y += 1
+    x = 0
+  result.toString
+
+def part1(input: String): Int =
+  val heightMap = Heightmap.fromString(input)
+
+  heightMap.lowPointsPositions.map(heightMap(_) + 1).sum
+end part1
+
+
+def bassin(lowPoint: Position, heightMap: Heightmap): Set[Position] =
+  @scala.annotation.tailrec
+  def iter(visited: Set[Position], toVisit: Queue[(Position, Height)], bassinAcc: Set[Position]): Set[Position] =
+    if toVisit.isEmpty then bassinAcc
+    else
+      val ((currentPos, currentValue), others) = toVisit.dequeue
+      val newNodes = heightMap.neighborsOf(currentPos).toList.filter { (pos, height) =>
+        !visited(currentPos) && height != 9 && currentValue < height 
```

Comment:
Given that basins are always separated by `9`'s (which seems weird to me, but the statement is quite clear about that), I don't think `currentValue < height` matters here. I believe it can be removed without losing correctness.

That means that it is not necessary to store `currentValue` at all, and that therefore a `Queue[Position]` would be enough.

Reply:
I tested your suggestion and you seem to be correct, although this is not intuitive for me.

You are right that `9` should not be included, however in my reasoning, I needed to include only elements that were bigger when *growing* the basin.
In other words `1 2 3 2`  I should stop at not include the `2` which led me to include this check.

I am a bit confused: do we extend until we reach a 9 no matter what value is encountered?

Reply:
The thing is that the statement rules out such situations, because it says:

> Locations of height 9 do not count as being in any basin, and all other locations will always be part of exactly one basin.

So the `3` would either have to part of the basin on the left, or on the right. But neither both or none. That means you could never know on which side it is supposed to go, and that therefore such situations are illegal.

Reply:
So when we consider all basins, all values will covered but 9.
It works looking at the example.
I am still puzzled but I understand better